test/test_in_toto_record.py

```diff
@@ -0,0 +1,185 @@
+    material. """
+    self.test_dir = tempfile.mkdtemp()
+    os.chdir(self.test_dir)
+
+    self.step_name = "test-step"
+    self.link_name_unfinished = ".{}.link-unfinished".format(self.step_name)
+
+    self.key_path = "test_key"
+    generate_and_write_rsa_keypair(self.key_path)
+
+    self.test_material= "test_material"
+    open(self.test_material, "w").close()
+
+  @classmethod
+  def tearDownClass(self):
+    """Change back to initial working dir and remove temp test directory. """
+    os.chdir(WORKING_DIR)
+    shutil.rmtree(self.test_dir)
+
+  def test_create_link_unfinished_file(self):
+    """Test record start creates a link metadata file. """
+    in_toto_record_start(self.step_name, self.key_path, [self.test_material])
+    open(self.link_name_unfinished)
+
+  def test_record_test_material(self):
+    """Test record start records expected material. """
+    in_toto_record_start(
+        self.step_name, self.key_path, [self.test_material])
+    link = Link.read_from_file(self.link_name_unfinished)
+    self.assertEquals(link.materials.keys(), [self.test_material])
+
+
+class TestInTotoRecordStop(unittest.TestCase):
+  """"Test in_toto_record_stop(step_name, key_path, product_list). """
+
+  @classmethod
+  def setUpClass(self):
+    """Create and change into temporary directory, generate two key pairs
+    and dummy product. """
+    self.test_dir = tempfile.mkdtemp()
+    os.chdir(self.test_dir)
+
+    self.step_name = "test-step"
+    self.link_name = "{}.link".format(self.step_name)
+    self.link_name_unfinished = ".{}.link-unfinished".format(self.step_name)
+
+    self.key_path = "test-key"
+    self.key_path2 = "test-key2"
+    generate_and_write_rsa_keypair(self.key_path)
+    generate_and_write_rsa_keypair(self.key_path2)
+
+    self.test_product = "test_product"
+    open(self.test_product, "w").close()
+
+  @classmethod
+  def tearDownClass(self):
+    """Change back to initial working dir and remove temp test directory. """
+    os.chdir(WORKING_DIR)
+    shutil.rmtree(self.test_dir)
+
+  def test_record_test_product(self):
+    """Test record stop records expected product. """
+    in_toto_record_start(self.step_name, self.key_path, [])
+    in_toto_record_stop(
+        self.step_name, self.key_path, [self.test_product])
+    link = Link.read_from_file(self.link_name)
+    self.assertEquals(link.products.keys(), [self.test_product])
+
+  def test_replace_unfinished_file(self):
+    """Test record stop removes unfinished file. """
+    in_toto_record_start(self.step_name, self.key_path, [])
+    in_toto_record_stop(self.step_name, self.key_path, [])
+    with self.assertRaises(IOError):
+      open(self.link_name_unfinished, "r")
+    open(self.link_name, "r")
+
+  def test_missing_unfinished_file(self):
+    """Test record stop exits on missing unfinished file. """
+    with self.assertRaises(SystemExit):
+      in_toto_record_stop(self.step_name, self.key_path, [])
```

Comment:
Is this test assuming that the tests run in order? that may not always be the case iirc. Maybe we want to ensure that there is no unfinished file before the `with self.assertRaises()`